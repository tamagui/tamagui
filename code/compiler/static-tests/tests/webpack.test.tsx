import '@expo/match-media'
import { render, screen } from '@testing-library/react'
import { describe, expect, test } from 'vitest'

/**
 * disabled for now but we really need to bring this back
 */

function getTest(name: string) {
  const app = require('./spec/out/out-webpack')
  const App = app[name]
  if (!App) {
    throw new Error(`No app found: ${name} in ${Object.keys(app || {})}`)
  }
  const Provider = app.Provider
  return {
    Element: (props: any) => (
      <Provider>
        <App {...props} />
      </Provider>
    ),
    renderTrue: () =>
      render(
        <Provider>
          <App conditional={true} />
        </Provider>
      ),
    renderFalse: () =>
      render(
        <Provider>
          <App conditional={false} />
        </Provider>
      ),
  }
}

describe('webpack-tests', () => {
  test('1. extracts to a div for simple views, flat transforms', () => {
    const { renderTrue } = getTest('Test1')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('2. extracts className for complex views but keeps other props', () => {
    const { renderTrue, renderFalse } = getTest('Test2')
    const { container: containerTrue } = renderTrue()
    expect(containerTrue).toMatchSnapshot()
    const { container: containerFalse } = renderFalse()
    expect(containerFalse).toMatchSnapshot()
  })

  test('5. spread conditional', () => {
    const { renderTrue } = getTest('Test5')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('6. spread ternary', () => {
    const { renderTrue, renderFalse } = getTest('Test6')
    expect(renderTrue().container).toMatchSnapshot()
    expect(renderFalse().container).toMatchSnapshot()
  })

  test('7. ternary + data-is', () => {
    const { renderTrue } = getTest('Test7')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('8. styleExpansions', () => {
    const { renderTrue } = getTest('Test8')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('9. combines with classname', () => {
    const { renderTrue } = getTest('Test9')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('10. extracts Text', () => {
    const { renderTrue } = getTest('Test10')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('12. ternary multiple on same key', () => {
    const { renderTrue } = getTest('Test12')
    expect(renderTrue().container).toMatchSnapshot()
  })

  test('14. extracts pseudo styles and evaluates constants', () => {
    const { renderTrue } = getTest('Test14')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('15. extracts spacer (complex expansion)', () => {
    const { renderTrue } = getTest('Test15')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('16. deopt when spreading multiple', () => {
    const { renderTrue } = getTest('Test16')
    const { container } = renderTrue()
    expect(container).toMatchSnapshot()
  })

  test('17. variant default false flattens properly', () => {
    const { renderTrue: renderOn } = getTest('TestVariantDefaultFalseOn')
    const { renderTrue: renderOff } = getTest('TestVariantDefaultFalseOff')

    const { container: containerOn } = renderOn()
    const { container: containerOff } = renderOff()

    const outCn = containerOn.firstChild?.firstChild?.['className']
    const out2Cn = containerOff.firstChild?.firstChild?.['className']

    expect(outCn).not.toContain(`_pl-t-space-4`)
    expect(out2Cn).toContain(`_pl-t-space-4`)
  })
})
