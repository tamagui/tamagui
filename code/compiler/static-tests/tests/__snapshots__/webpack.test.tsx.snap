// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`webpack-tests > 1. extracts to a div for simple views, flat transforms 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="  _dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-1 _fd-column _fg-1 _btlr-100px _btrr-100px _bbrr-100px _bblr-100px _bg-red _bxsh-0px0px10pxr538535358 _transform-0hover-scale21281  test1"
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color font_body "
      >
        hello world
      </span>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 2. extracts className for complex views but keeps other props 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="  _dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-row _ox-hidden _oy-hidden  who"
    />
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-row _pt-2px _pb-2px _t-0px ease-in-out-top _t--14px _bg-fff35 "
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color font_body "
      >
        hello world
      </span>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 2. extracts className for complex views but keeps other props 2`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="  _dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-row _ox-hidden _oy-hidden  who"
    />
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-row _bg-00035 _pt-2px _pb-2px _t-0px ease-in-out-top "
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color font_body "
      >
        hello world
      </span>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 5. spread conditional 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _ox-hidden _oy-hidden _bg-blue "
    >
      <div
        class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _ox-hidden _oy-hidden hello-world _bg-blue "
      />
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 6. spread ternary 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _ox-hidden _oy-hidden _bg-blue "
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color font_body "
      >
        hello world
      </span>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 6. spread ternary 2`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _ox-hidden _oy-hidden _bg-red "
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color font_body "
      >
        hello world
      </span>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 7. ternary + data-is 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _fs-0 _fd-column _pr-11px _pl-11px _pb-10px _w-6637 _miw-500px _maw-3037"
    >
      <div
        class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _w-10px _h-10px "
      />
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 8. styleExpansions 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="  _dsp-flex _ai-stretch _fb-auto _bxs-border-box _mih-0px _miw-0px _fs-0 _fd-column _pos-relative _t-0px _l-0px _r-0px _b-0px "
    >
      <div
        class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _bg-red"
      >
        <div
          class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _mih-0px _miw-0px _fs-0 _fd-column _pos-relative _t-0px _l-0px _r-0px _b-0px _bg-white _col-blue"
        />
      </div>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 9. combines with classname 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="  _dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _pt-15px _pb-15px  home-top-dish"
    />
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 10. extracts Text 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <span
      class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color _fos-10px font_body "
    >
      <span
        class="  _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-color _ussel-auto font_body "
      >
        <span
          class="font_body _ff-f-family _dsp-inline _bxs-border-box _ww-break-word _whiteSpace-pre-wrap _mt-0px _mr-0px _mb-0px _ml-0px _col-00035 _fos-15px _fow-600"
        >
          hello
        </span>
      </span>
    </span>
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 12. ternary multiple on same key 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column _o-1 _transform- "
    />
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 14. extracts pseudo styles and evaluates constants 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_bg-0hover-red _bg-0active-blue _dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-0 _fd-column"
    />
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 15. extracts spacer (complex expansion) 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <span
      class="  is_Spacer _dsp-flex _ai-stretch _fd-column _fb-auto _bxs-border-box _pos-relative _mih-t-space-tru101 _miw-t-space-tru101 _fs-0 _w-t-space-tru101 _h-t-space-tru101 _pe-none "
    />
    <span
      class="  is_Spacer _dsp-flex _ai-stretch _fd-column _fb-auto _bxs-border-box _pos-relative _mih-10px _miw-10px _fs-1 _pe-none _fg-1 _w-10px _h-10px "
    />
    <div
      style="display: contents;"
    />
  </span>
</div>
`;

exports[`webpack-tests > 16. deopt when spreading multiple 1`] = `
<div>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_dsp-flex _ai-stretch _fb-auto _bxs-border-box _pos-relative _mih-0px _miw-0px _fs-1 _fd-column _fg-1 _bg-red"
    />
    <div
      style="display: contents;"
    />
  </span>
</div>
`;
